### ISSUE TITLE
Estimation of barometric pressure/altitude relationship?

### ISSUE THREAD

**Issue by dmcclean** [GaloisInc/estimator#3]
I was curious to see that the `StateVector` type in `Numeric.Estimator.Model.SensorFusion` doesn't include estimation of the local relationship between barometric pressure and altitude. (Or, maybe it does, but it's implicit and I overlooked it?)

That state, [QNH](http://en.wikipedia.org/wiki/QNH), is disseminated by voice in piloted flight, but it can quickly be inferred from GPS altitudes and barometer readings. It seems like it would be useful for small UAV flight too, since it would allow you to ride through lengthy GPS signal interruptions without losing absolute altitude information? (Perhaps you could pick a relatively safe horizontal navigation strategy like circling, hovering, or climbing above some mapped equivalent of the [minimum vectoring altitude](http://www.skybrary.aero/index.php/Minimum_Vectoring_Altitude_%28MVA%29), and you'd have a decent chance of avoiding CFIT even with large horizontal position variances.

**Comment by jameysharp** [comment]
You're right, the sensor fusion model doesn't yet estimate that state. `Numeric.Estimator.Model.Pressure` currently just uses the 1976 U.S. Standard Atmosphere model as specified from sea level to 11km.

I think that's a pretty straightforward change and would love to see a pull request implementing it!

**Comment by dmcclean** [comment]
OK, great, will do. I just like to ask first on projects that haven't seen many outside contributions. Also I wasn't sure if it was happening somewhere and I just wasn't seeing it. :)

I'm not sure what your policy on dependencies is, but are you aware of the [atmos](http://hackage.haskell.org/package/atmos) package? It does the entire 1976 US Standard Atmosphere to 86km and with a good approximation above that. I helped contribute to adding the inverse pressure-to-height ("pressure altitude") function, so now it also does that.

**Comment by dmcclean** [comment]
Do you have any sample client code that I can imitate? I'm having trouble with a few things.

Most notably with getting the type of `SensorFusion.processModel` to match the argument of `EKFProcess`. I think it has something to do with the `augmentProcess` function, but the dominoes aren't lining up for me.

I also don't quite follow the situation with `DisturbanceVector`. I would've expected the things in that vector to be control inputs to the process, like what body rates the controller thought it was achieving by moving control surfaces, or something along those lines. Is this the case and the name `disturbanceGyro` isn't really associated with the gyro at all? Or is something else going on? Does it work at all to null out these disturbance inputs and increase the process covariance if the filter is being employed just for tracking and it doesn't know what maneuvers are being attempted?

**Comment by pchickey** [comment]
Our client code is found in a separate repo -
https://github.com/GaloisInc/smaccmpilot-stm32f4/tree/master/src/smaccm-ins

Note that it depends heavily on other Ivory lang infrastructure, one
easy way to get all the dependencies you need is

git clone --recursive --branch feature/tower9
https://github.com/galoisinc/smaccmpilot-build

and then advance the `smaccmpilot-stm32f4` submodule to the
master branch.

Jamey is still actively developing the smaccm-ins code and we don't yet
have a production ready version for our target platforms, the PX4FMU 1.7
and Pixhawk flight controllers from the PX4 Autopilot project
(http://pixhawk.org/). However, we expect to get the final integration
issues worked out in the next couple of weeks.

On Sat, Jan 31, 2015, at 03:22 PM, Douglas McClean wrote:
> Do you have any sample client code that I can imitate? I'm having
> trouble with a few things.


> Most notably with getting the type of SensorFusion.processModel to
> match the argument of EKFProcess. I think it has something to do with
> the augmentProcess function, but the dominoes aren't lining up for me.


> I also don't quite follow the situation with DisturbanceVector. I
> would've expected the things in that vector to be control inputs to
> the process, like what body rates the controller thought it was
> achieving by moving control surfaces, or something along those lines.
> Is this the case and the name disturbanceGyro isn't really associated
> with the gyro at all? Or is something else going on? Does it work at
> all to null out these disturbance inputs and increase the process
> covariance if the filter is being employed just for tracking and it
> doesn't know what maneuvers are being attempted?


> — Reply to this email directly or view it on GitHub[1].

>



Links:

  1. https://github.com/GaloisInc/estimator/issues/3#issuecomment-72342267

**Comment by jameysharp** [comment]
I hope Pat's pointers were some help. To be more specific, take a look at [SMACCMPilot.INS.Simulate](https://github.com/GaloisInc/smaccmpilot-stm32f4/blob/master/src/smaccm-ins/src/SMACCMPilot/INS/Simulate.hs). It ties this sensor fusion model to a State monad over KalmanState.

The DisturbanceVector isn't being used to represent control inputs here. Instead, the idea is to drive the process model from measured acceleration and rotation rates. I wish I had a paper or book to cite that describes this technique, but so far as far as I've been able to tell, it's just sensor fusion folklore. As a result, I can't say definitively why people do this, but I think there are a couple of good reasons:

- The state vector doesn't need to include acceleration or rotation rate, because they aren't included in any measurement models and because the process model can't do better than assuming they haven't changed. This reduces the dimensionality of the state, which generally improves both the computation cost and the filter's ability to produce good results.

- If you do include acceleration and rotation rate in the state vector, then you need to add process noise for them that is large enough to cover all possible jerks and torques that your vehicle might experience. This may not be too bad for a quadcopter, but I've also tried to do it for an amateur rocket, which can go from 0 to 100m/s^2 in tens of milliseconds. In that setting, constant process noise causes the covariance to blow up to unusable levels very quickly; and trying to adjust the process noise for different flight phases was, in my experience, unreliable.

This was an excellent question and probably calls for better documentation.

tl;dr: Fill your accelerometer and gyro measurements into the DisturbanceVector. :-)

**Comment by dmcclean** [comment]
Ah, thanks very much for that extra explanation. I am still way too new at this to have read that between the lines, but that is a clever trick that makes a lot of sense, and seems like a great idea for the application for the reasons you mentioned. Pat's notes were helpful but I haven't quite gotten the whole thing yet.

I assume you then treat the measurement noise of your IMU sensors as velocity/orientation process noise (except that the dimensions don't match, so you somehow factor in the sampling time)? I take that to be the meaning of [this comment](https://github.com/GaloisInc/estimator/blob/master/src/Numeric/Estimator/Model/SensorFusion.hs#L126-L129), but I'm not quite sure, partly because I am having trouble tracking down where the process noise is defined. I can find the initial state covariance, but looking [here](https://github.com/GaloisInc/estimator/blob/master/src/Numeric/Estimator/KalmanFilter.hs#L80-L89), I don't see where the process noise is added to the a priori state covariance estimate?

**Comment by jameysharp** [comment]
Process noise is added in the [`Process` instance](https://github.com/GaloisInc/estimator/blob/master/src/Numeric/Estimator/KalmanFilter.hs#L59), although augmenting the process model with measured disturbances makes it more confusing. Have a look at the [`runProcessModel` driver in `Simulate.hs`](https://github.com/GaloisInc/smaccmpilot-stm32f4/blob/master/src/smaccm-ins/src/SMACCMPilot/INS/Simulate.hs#L34). There you'll see an example of `augmentProcess` taking two separate process covariance matrices: one for the portion of the process which applies to the real state vector, and another for the portion operating on the disturbance vector.

This notion of an augmented process model, where part of the state vector and covariance is virtual rather than persisting across timesteps, also warrants more explanation, although I /did/ write Haddock for Numeric.Estimator.Augment, which I hope will help. Again I can't provide a good citation for this, this time because I invented it to make the math match the implementation in Paul Riseborough's [InertialNav](https://github.com/priseborough/InertialNav) project without having to special-case the EKF generator. I assume the technique must be known in the literature somewhere...

By the way, I have no problem answering these questions here, but just want to reiterate that the original topic of this particular issue was about estimating the barometric pressure offset relative to the standard atmosphere model. If we come up with other worthwhile changes, I'd like those in their own issues.

**Comment by dmcclean** [comment]
Ah, OK, there it is, thanks.

Yes, I apologize for the bad manners wrt cluttering up the issues. I was asking here as something of a last resort, because I wasn't getting very far figuring out how to get running so that I could test my changes without this sort of "overview" information. I couldn't find anywhere else to ask, and AFAIK there isn't any sort of private messaging between github accounts. :/

**Comment by jameysharp** [comment]
A quick follow-up: I'm still interested in this proposal and happy to answer questions, whenever you have time. :-)
